### PR TITLE
feat(udf): add udf metrics to embeded udf, refactor external udf metrics.

### PR DESCRIPTION
**Pull request by yufansong**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

1. Extract the UDF metrics into a higher layer.
2. Then metrics can work for every kinds of the udf
3. The external UDF call logic is little complicated. But we decide to count each trigger as only one call. Event it actually do several call, we can still infer this info from the latency metrics.
4. For embeded udf call, the `link` label will be empty ``. Show in following example, the first three record udf is external udf, the last one are embeded udf:
5. Also add `language` labels: `wasm`, `javascript`, `python`, `external`, for example, the first three is `external`, the last one is `javascript`
image

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

> we decide to count each trigger as only one call. Event it actually do several call, we can still infer this info from the latency metrics.

Actually I am a little worried about one case. When `always_retry_on_network_error` is enabled, the call may be retrying forever. In this case we will no longer see any UDF metrics, although error messages can still be seen in the log. Not sure if this would have any impact on diagnosing UDF problems. cc @kwannoel

**Comment by kwannoel** [comment]
> LGTM!
> 
> > we decide to count each trigger as only one call. Event it actually do several call, we can still infer this info from the latency metrics.
> 
> Actually I am a little worried about one case. When `always_retry_on_network_error` is enabled, the call may be retrying forever. In this case we will no longer see any UDF metrics, although error messages can still be seen in the log. Not sure if this would have any impact on diagnosing UDF problems. cc @kwannoel

Hmm could we have a way to track in-flight UDF calls? So if the number stays high for a long time, we know the udf server has issues.

File: src/expr/udf/src/lib.rs [hunk 1]
```diff
@@ -17,7 +17,8 @@
 
 mod error;
 mod external;
-mod metrics;
+pub mod metrics;
```
[the remark this answers is not in the document]
Comment (yufansong) [reply src/expr/udf/src/lib.rs:20]: Let me hold this first, I am thinking if we add some specific for in-flight udf, the metrics mod should better stay here. After solve this problem, I can move it in other PR.

**Comment by yufansong** [comment]
> > LGTM!
> > > we decide to count each trigger as only one call. Event it actually do several call, we can still infer this info from the latency metrics.
> > 
> > 
> > Actually I am a little worried about one case. When `always_retry_on_network_error` is enabled, the call may be retrying forever. In this case we will no longer see any UDF metrics, although error messages can still be seen in the log. Not sure if this would have any impact on diagnosing UDF problems. cc @kwannoel
> 
> Hmm could we have a way to track in-flight UDF calls? So if the number stays high for a long time, we know the udf server has issues.

Maybe I can add only udf failure count for `always_retry_on_network_error`, then we can see the udf count of failure

**Pull request merged**